Docs: Add documentation for library elements API (#39829)
grafana/grafana: docs/sources/http_api/library_element.md

The uid can have a maximum length of 40 characters.

## Get all library elements

`GET /api/library-elements`

Returns all library elements that the authenticated user has permission to view. You can control the maximum number of library elements returned through the `perPage` query parameter, the default is 100. You can also pass the `page` query parameter for fetching library elements from a page other than the first one.

Query parameters:

- **searchString** – Part of the name or description searched for
- **kind** – Kind of element to search for, `1` for library panels or `2` for library variables
- **sortDirection** – Sort order of elements, `alpha-asc` for ascending or `alpha-desc` for descending
- **typeFilter** – Comma separated list of types to filter elements by
- **excludeUid** – Element UID to exclude from results
- **folderFilter** – Comma separated list of folder id's to filter elements by
- **perPage** – Limit the number of returned results per page (default 100)

> - **page** – Use this parameter to access hits beyond `perPage`. Numbering starts at 1. `perPage` param acts as page size.

Author: I think this part is important `Numbering starts at 1. 'perPage' param acts as page size.`

> - **page** – Use this parameter to access hits beyond `perPage`. Numbering starts at 1. `perPage` param acts as page size.

Reviewer: Suggestion:
The page for a set of records, given that only `perPage` records are returned at a time. Numbering starts at 1.

     "totalCount": 15,
     "page": 1,
     "perPage": 10
     "elements": [
        {
            "id": 25,
            "orgId": 1,
            "folderId": 0,
            "uid": "V--OrYHnz",
            "name": "API docs Example",
            "kind": 1,
            "type": "text",
            "description": "",
            "model": {...},
            "version": 1,
            "meta": {
                "folderName": "General",
                "folderUid": "",
                "connectedDashboards": 1,
                "created": "2021-09-27T09:56:17+02:00",
                "updated": "2021-09-27T09:56:17+02:00",
                "createdBy": {
                    "id": 1,
                    "name": "admin",
                    "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
                },
                "updatedBy": {
                    "id": 1,
                    "name": "admin",
                    "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
                }
            }
        },
        {...}
        {...}
     ],
  }
}
```

Status Codes:

- **200** – Found
- **401** – Unauthorized

## Get library element by uid

`GET /api/library-elements/:uid`

> Will return the library-element given the uid.

Author: This endpoint returns a library element with the supplied UID.

Accept: application/json
Content-Type: application/json
Authorization: Bearer eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk
```

**Example Response**:

```http
HTTP/1.1 200
Content-Type: application/json

{
    "result": {
      "id": 25,
      "orgId": 1,
      "folderId": 0,
      "uid": "V--OrYHnz",
      "name": "API docs Example",
      "kind": 1,
      "type": "text",
      "description": "",
      "model": {...},
      "version": 1,
      "meta": {
          "folderName": "General",
          "folderUid": "",
          "connectedDashboards": 1,
          "created": "2021-09-27T09:56:17+02:00",
          "updated": "2021-09-27T09:56:17+02:00",
          "createdBy": {
              "id": 1,
              "name": "admin",
              "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
          },
          "updatedBy": {
              "id": 1,
              "name": "admin",
              "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
          }
      }
   }
}
```

Status Codes:

- **200** – Found
- **401** – Unauthorized
- **404** – Library element not found

## Get library element by name

`GET /api/library-elements/name/:name`

> Will return the library-element given the name.

Author: This endpoint returns a library element with the given name.

Accept: application/json
Content-Type: application/json
Authorization: Bearer eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk
```

**Example Response**:

```http
HTTP/1.1 200
Content-Type: application/json

{
    "result": {
      "id": 25,
      "orgId": 1,
      "folderId": 0,
      "uid": "V--OrYHnz",
      "name": "API docs Example",
      "kind": 1,
      "type": "text",
      "description": "",
      "model": {...},
      "version": 1,
      "meta": {
          "folderName": "General",
          "folderUid": "",
          "connectedDashboards": 1,
          "created": "2021-09-27T09:56:17+02:00",
          "updated": "2021-09-27T09:56:17+02:00",
          "createdBy": {
              "id": 1,
              "name": "admin",
              "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
          },
          "updatedBy": {
              "id": 1,
              "name": "admin",
              "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
          }
      }
   }
}
```

Status Codes:

- **200** – Found
- **401** – Unauthorized
- **404** – Library element not found

## Get library element connections

`GET /api/library-elements/:uid/connections`

> Will return a list of connections for a library-element given the uid.

Author: I think UID is an important piece of information in this context.

**Example Request**:

```http
GET /api/library-elements/V--OrYHnz/connections HTTP/1.1
Accept: application/json
Content-Type: application/json
Authorization: Bearer eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk
```

**Example Response**:

```http
HTTP/1.1 200
Content-Type: application/json

{
    "result": [
        {
            "id": 148,
            "kind": 1,
            "elementId": 25,
            "connectionId": 527,
            "created": "2021-09-27T10:00:07+02:00",
            "createdBy": {
                "id": 1,
                "name": "admin",
                "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
            }
        }
    ]
}
```

Status Codes:

- **200** – Found
- **401** – Unauthorized
- **404** – Library element not found

## Create library element

`POST /api/library-elements`

Creates a new library element.

JSON Body schema:

- **folderId** – The folder id of the folder where the library element is stored.
- **name** – The name of the library element.
- **model** – The json model for the library element.
- **kind** – Kind of element to create, `1` for library panels or `2` for library variables.

> - **uid** – Optional [unique identifier](/http_api/library_element/#identifier-id-vs-unique-identifier-uid).

Reviewer: Make similar changes suggested in other review comments.

Content-Type: application/json
Authorization: Bearer eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk

{
  "uid": "nErXDvCkzz",
  "folderId": 0,
  "name": "Example library panel",
  "model": {...},
  "kind": 1
}
```

**Example Response**:

```http
HTTP/1.1 200
Content-Type: application/json

{
    "result": {
        "id": 28,
        "orgId": 1,
        "folderId": 0,
        "uid": "nErXDvCkzz",
        "name": "Example library panel",
        "kind": 1,
        "type": "",
        "description": "",
        "model": {...},
        "version": 1,
        "meta": {
            "folderName": "General",
            "folderUid": "",
            "connectedDashboards": 0,
            "created": "2021-09-30T09:14:22.378307+02:00",
            "updated": "2021-09-30T09:14:22.378307+02:00",
            "createdBy": {
                "id": 1,
                "name": "admin",
                "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
            },
            "updatedBy": {
                "id": 1,
                "name": "admin",
                "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
            }
        }
    }
}
```

Status Codes:

- **200** – Created

> - **400** – Errors (name or uid already exists, invalid json, missing or invalid fields, etc.)

Reviewer: ```suggestion
- **400** – Errors (for example, name or UID already exists, invalid JSON, missing or invalid fields, and so on).
```

- **401** – Unauthorized

> - **403** – Access Denied

Reviewer: ```suggestion
- **403** – Access denied
```

## Update library element

`PATCH /api/library-elements/:uid`

Updates an existing library element identified by uid.

JSON Body schema:

> - **folderId** – The folder id of the folder where the library element is stored.

Reviewer: ```suggestion
- **folderId** – ID of the folder where the library element is stored.
```

> - **name** – The name of the library element.

Reviewer: ```suggestion
- **name** –Name of the library element.
```

> - **model** – The json model for the library element.

Reviewer: ```suggestion
- **model** – The JSON model for the library element.
```

> - **kind** – Kind of element to create, `1` for library panels or `2` for library variables.

Reviewer: ```suggestion
- **kind** – Kind of element to create. Use `1` for library panels or `2` for library variables.
```

> - **version** – Version of the library element being updated.

Reviewer: ```suggestion
- **version** – Version of the library element you are updating.
```

> - **uid** – Optional [unique identifier](/http_api/library_element/#identifier-id-vs-unique-identifier-uid).

Reviewer: ```suggestion
- **uid** – Optional, the [unique identifier](/http_api/library_element/#identifier-id-vs-unique-identifier-uid).
```

Authorization: Bearer eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk

{
  "name": "Renamed library panel",
  "kind": 1,
  "version": 1
}
```

**Example Response**:

```http
HTTP/1.1 200
Content-Type: application/json

{
    "result": {
        "id": 28,
        "orgId": 1,
        "folderId": 0,
        "uid": "nErXDvCkzz",
        "name": "Renamed library panel",
        "kind": 1,
        "type": "",
        "description": "",
        "model": {
            "description": "",
            "type": ""
        },
        "version": 2,
        "meta": {
            "folderName": "General",
            "folderUid": "",
            "connectedDashboards": 0,
            "created": "2021-09-30T09:14:22+02:00",
            "updated": "2021-09-30T09:25:57.697214+02:00",
            "createdBy": {
                "id": 1,
                "name": "admin",
                "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
            },
            "updatedBy": {
                "id": 1,
                "name": "admin",
                "avatarUrl": "/avatar/46d229b033af06a191ff2267bca9ae56"
            }
        }
    }
}
```

Status Codes:

- **200** – Updated

> - **400** – Errors (name or uid already exists, invalid json, missing or invalid fields, etc.)

Reviewer: ```suggestion
- **400** – Errors (for example, name or UID already exists, invalid JSON, missing or invalid fields, and so on).
```

- **401** – Unauthorized

> - **403** – Access Denied

Reviewer: ```suggestion
- **403** – Access denied
```

- **404** – Library element not found
- **412** – Version mismatch

## Delete library element

`DELETE /api/library-elements/:uid`

> Deletes an existing library element identified by UID. Will not succeed if the library element is connected. This operation cannot be reverted.

Reviewer: ```suggestion
Deletes an existing library element as specified by the UID. This operation cannot be reverted.

>**Note:** You cannot delete a library element that is connected. This operation cannot be reverted.
```

The maximum length of a UID is 40 characters.

## Get all library elements

`GET /api/library-elements`

Returns a list of all library elements the authenticated user has permission to view. Use `perPage` query parameter to control the maximum number of library elements returned; the default limit is 100. You can also use the `page` query parameter to fetch library elements from any page other than the first one.

Query parameters:

- **searchString** – Part of the name or description searched for
- **kind** – Kind of element to search for. Use `1` for library panels or `2` for library variables.
- **sortDirection** – Sort order of elements. Use `alpha-asc` for ascending and `alpha-desc` for descending sort order.
- **typeFilter** – A comma separated list of types to filter the elements by.
- **excludeUid** – Element UID to exclude from search results.
- **folderFilter** – A comma separated list of folder ID(s) to filter the elements by.

> - **perPage** – Limit the number of results per page; default is 100.

Reviewer: ```suggestion
- **perPage** – The number of results per page; default is 100.
```
